### ISSUE TITLE
Source reservations and shipment flow

### ISSUE THREAD

**Issue by paales** [magento/inventory#1441]
_Note: This is discussed earlier during the MSI Open Grooming meeting (https://www.youtube.com/watch?v=fl8anqyN-1Q&index=2&list=PLrQ5FBCRsEbWKK6U_3Awe7X-nG7KY0WPW) but after the meeting there were some questions how certain specifics were implemented._

### Backstory
We have a feature request from a customer to implement purchase orders and therefor it lead me to the idea that we should have some form of source reservations. But while writing the specifications it came to my attention that we have a feature-gap in the new MSI flow regarding shipments.

### Problem
We need to make an API-call to the actual source (warehouse) to actually request as shipment for certain products. But to know what to actually communicate to the warehouse, the Source Selection Algorithm needs to have run. We can't run that algorithm during the shipment creation because the product hasn't actually shipped and thus the source isn't deducted at that point.

### Proposed solution
We split the Source Selection Algorithm and the Shipment creation into two steps, while making sure the source, stock and salable qty remain consistent.

#### Source Selection Algorithm step
The orders' stock reservation is nullified, source reservation is made.

_After this step the Source Selection Algorithm has ran and the sources are selected. This can be used as a starting point to communicate the shipment to the source warehouse._

#### Shipment step
The orders' source reservation is nullified, actual source deduction is made by the shipment.

### Implementation
One of the questions is that if we split the SSA from the shipment, when will we run the SSA? Because in the manual-order-processing case it doesn't need to be run at all, it can happen exactly the same way it happens now, but in the API-case it needs to be ran as soon as possible.

Proposal: Always run the default SSA (async) after an order has been placed and store the result in the source reservation table.

The value of the result of the SSA can be used as prefilled values that can be used when the UI-part has ran. If the result there is different, the current reservations will be reverted and a new reservation is made.

**Comment by aleron75** [comment]
>Always run the default SSA (async) after an order has been placed and store the result in the source reservation table.

SSA gives a result that can be changed by the merchant while processing an order.

Automatically applying the default SSA asynchronously would imply accepting the result as-is. 
The merchant should be able to enable or disable this behavior.

This is not the feature that makes me like the idea of source-level reservations.

What I like of source-level reservations is the fact that they seem to **better represent the domain**.

Once it's defined where products have to be picked from (sources), creating source-level reservations would correspond to what happens in reality: products are phisically or virtually picked (reserved) from the source but they still have to be shipped. 

Shipping is something that happens only after products are packed and assigned to a logistic service.

That said, I agree with @maghamed that this is a complex architectural change and I wouldn't recommend to do it at this stage of development.

**Comment by paales** [comment]
> This is not the feature that makes me like the idea of source-level reservations.

Ok, what is your proposal then, when should the SSA be ran? I think _asap_ is a good moment, because the actual reservation needs to be made as soon as possible in the real warehouse..

> Shipping is something that happens only after products are packed and assigned to a logistic service.

I understand what you guys want the shipment to be, therefor I tried to create an alternative that is a bit more conservative from my point of view.

> That said, I agree with @maghamed that this is a complex architectural change and I wouldn't recommend to do it at this stage of development.

I wasn't aware that he said that, but let me explain. I'm not looking to rush this for a 2.3 release if that undermines quality etc. If you guys determine that it is unsuitable to integrate this right because of the scope of the change, thats fine, we can integrate this with 2.4 (2.3.x is probably out of the question because of the potentional API change).

**Comment by paales** [comment]
We had a discussion on Slack, but before the messages get lost:

There are two main questions that I’d like to get answered:
- If source reservations is enabled: When should we run the SSA? (made a proposal here https://github.com/magento-engcom/msi/issues/1441)
- If source reservations is enabled: Are we going with the ‘pending shipments’ or just bare source reservations.

**Response by @maghamed written on June 29th:**

> If source reservations is enabled: When should we run the SSA?

Currently we have next steps via API to process the Order:

1) Most of the steps are equivalent to typical Order placement/processing https://devdocs.magento.com/guides/v2.2/get-started/order-tutorial/order-create-shipment.html

2) But before shipment step integrator has to call additional one (located in InventorySourceSelectionApi)

```xml
<route url="/V1/inventory/source-selection-algorithm-result" method="POST">
    <service class="Magento\InventorySourceSelectionApi\Api\SourceSelectionServiceInterface" method="execute"/>
    <resources>
        <resource ref="Magento_InventoryApi::source"/>
    </resources>
</route>
``` 
this service would provide a suggestion how better to fulfill given order and from which sources to make a deduction.

3)  After that creating dedicated shipments (each per specified Source) merchant should specify an extension attribute `source_code` for ShipmentInterface

```xml
<config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:framework:Api/etc/extension_attributes.xsd">
    <extension_attributes for="Magento\Sales\Api\Data\ShipmentInterface">
        <attribute code="source_code" type="string" />
    </extension_attributes>
</config>
```

Doing so, merchant specify and persist source deduction from particular source (edited)
This is how we changed default scenario working via WEB API for integrator
How do you propose to change this introducing Source level reservations?

Looks like we need to introduce additional call after step `2` (and before `3`) to create Source reservations. And release these reservations the shipment (step `3`) is made

**Comment by paales** [comment]
We've implemented this in a module that is currently running in production:
https://www.reachdigital.nl/en/magento-module-development/magento-2-order-source-reservation-module

If you're interested, please let me know.

**Comment by sidolov** [comment]
Hi @paales , we are about to complete MSI project feature development and keep issues of defined scope only to have clean and transparent backlog. 
Unfortunately, the current issue is out of scope and should be closed. 
Thank you for your involvement!
